buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java

```diff
@@ -538,30 +538,38 @@ static void setBytes(AbstractByteBuf buf, long addr, int index, byte[] src, int
     }
 
     static void setBytes(AbstractByteBuf buf, long addr, int index, ByteBuffer src) {
-        buf.checkIndex(index, src.remaining());
-
-        int length = src.remaining();
+        final int length = src.remaining();
         if (length == 0) {
             return;
         }
 
         if (src.isDirect()) {
+            buf.checkIndex(index, length);
             // Copy from direct memory
             long srcAddress = PlatformDependent.directBufferAddress(src);
-            PlatformDependent.copyMemory(srcAddress + src.position(), addr, src.remaining());
+            PlatformDependent.copyMemory(srcAddress + src.position(), addr, length);
             src.position(src.position() + length);
         } else if (src.hasArray()) {
+            buf.checkIndex(index, length);
             // Copy from array
             PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), addr, length);
             src.position(src.position() + length);
         } else {
-            ByteBuf tmpBuf = buf.alloc().heapBuffer(length);
-            try {
-                byte[] tmp = tmpBuf.array();
-                src.get(tmp, tmpBuf.arrayOffset(), length); // moves the src position too
-                PlatformDependent.copyMemory(tmp, tmpBuf.arrayOffset(), addr, length);
-            } finally {
-                tmpBuf.release();
+            if (length < 8) {
+                buf.checkIndex(index, length);
+                final int srcPosition = src.position();
+                int srcIndex = srcPosition;
+                long dstAddr = addr;
+                for (int i = 0; i < length; i++) {
```

Comment:
You could also just use the srcIndex to calculate how often you need to loop. This way you would safe one increment op.